Fix Issue #3325 (#5004)
All-Hands-AI/OpenHands: docs/modules/usage/configuration-options.md

---
id: configuration-options
title: Configuration Options
sidebar_label: Configuration Options
---

# Configuration Options

> This guide details all configuration options available for OpenHands, helping you customize its behavior and integrate it with other services. Required options and those with recommended settings are highlighted.

Reviewer: ```suggestion
This guide details all configuration options available for OpenHands, helping you customize its behavior and integrate it with other services.
```

   - [Dependencies and Environment](#dependencies-and-environment)
   - [Evaluation](#evaluation)
5. [Security Configuration](#security-configuration)
   - [Confirmation Mode](#confirmation-mode)
   - [Security Analyzer](#security-analyzer)

---

## Core Configuration
The core configuration options are defined in the `[core]` section of the `config.toml` file.

**API Keys**
- `e2b_api_key`
  - Type: `str`
  - Default: `""`
  - Description: API key for E2B

- `modal_api_token_id`
  - Type: `str` 
  - Default: `""`
  - Description: API token ID for Modal

- `modal_api_token_secret`
  - Type: `str`
  - Default: `""`
  - Description: API token secret for Modal

**Workspace**
- `workspace_base`
  - Type: `str`
  - Default: `"./workspace"`
  - Description: Base path for the workspace

- `cache_dir`
  - Type: `str`
  - Default: `"/tmp/cache"`
  - Description: Cache directory path

**Debugging and Logging**
- `debug`
  - Type: `bool`
  - Default: `false`
  - Description: Enable debugging

- `disable_color`
  - Type: `bool`
  - Default: `false`
  - Description: Disable color in terminal output

**Session Management**
- `enable_cli_session`
  - Type: `bool`
  - Default: `false`

>   - Description: Enable saving and restoring the session when run from CLI

Reviewer: Just a small thing here, this does not exist anymore. It was removed sometime last month or so

**Trajectories**
- `trajectories_path`
  - Type: `str`
  - Default: `"./trajectories"`

>   - Description: Path to store trajectories (can be a folder or a file)

Reviewer: ```suggestion
  - Description: Path to store trajectories (can be a folder or a file). If it's a folder, the trajectories will be saved in a file named with the session id name and .json extension, in that folder.
```

  - Type: `str`
  - Default: `"memory"`
  - Description: File store type

- `file_uploads_allowed_extensions`
  - Type: `list of str`
  - Default: `[".*"]`
  - Description: List of allowed file extensions for uploads

- `file_uploads_max_file_size_mb`
  - Type: `int`
  - Default: `0`
  - Description: Maximum file size for uploads, in megabytes

- `file_uploads_restrict_file_types`
  - Type: `bool`
  - Default: `false`
  - Description: Restrict file types for file uploads

- `file_uploads_allowed_extensions`
  - Type: `list of str`
  - Default: `[".*"]`
  - Description: List of allowed file extensions for uploads

**Task Management**
- `max_budget_per_task`
  - Type: `float`
  - Default: `0.0`
  - Description: Maximum budget per task (0.0 means no limit)

- `max_iterations`
  - Type: `int`
  - Default: `100`
  - Description: Maximum number of iterations

**Sandbox Configuration**
- `workspace_mount_path_in_sandbox`
  - Type: `str`
  - Default: `"/workspace"`
  - Description: Path to mount the workspace in the sandbox

- `workspace_mount_path`
  - Type: `str`
  - Default: `""`
  - Description: Path to mount the workspace

- `workspace_mount_rewrite`
  - Type: `str`
  - Default: `""`

>   - Description: Path to rewrite the workspace mount path to

Reviewer: ```suggestion
  - Description: Path to rewrite the workspace mount path to. You can usually ignore this, it refers to special cases of running inside another container.
```

**Miscellaneous**
- `run_as_openhands`
  - Type: `bool`
  - Default: `true`
  - Description: Run as OpenHands

- `runtime`
  - Type: `str`
  - Default: `"eventstream"`
  - Description: Runtime environment

- `default_agent`
  - Type: `str`
  - Default: `"CodeActAgent"`
  - Description: Name of the default agent

- `jwt_secret`
  - Type: `str`

>   - Default: `""`

Reviewer: ```suggestion
  - Default: `uuid.uuid4().hex`
```

>   - Description: JWT secret for authentication

Reviewer: ```suggestion
  - Description: JWT secret for authentication. Please set it to your own value.
```

## LLM Configuration
The LLM (Large Language Model) configuration options are defined in the `[llm]` section of the `config.toml` file.

**AWS Credentials**
- `aws_access_key_id`
  - Type: `str`
  - Default: `""`
  - Description: AWS access key ID

- `aws_region_name`
  - Type: `str`
  - Default: `""`
  - Description: AWS region name

- `aws_secret_access_key`
  - Type: `str`
  - Default: `""`
  - Description: AWS secret access key

**API Configuration**
- `api_key`
  - Type: `str`

>   - Default: `"your-api-key"`

Reviewer: ```suggestion
  - Default: `None`
```

  - Description: API key to use

- `base_url`
  - Type: `str`
  - Default: `""`
  - Description: API base URL

- `api_version`
  - Type: `str`
  - Default: `""`
  - Description: API version

- `input_cost_per_token`
  - Type: `float`
  - Default: `0.0`
  - Description: Cost per input token

- `output_cost_per_token`
  - Type: `float`
  - Default: `0.0`
  - Description: Cost per output token

**Custom LLM Provider**
- `custom_llm_provider`
  - Type: `str`
  - Default: `""`
  - Description: Custom LLM provider

**Embeddings**
- `embedding_base_url`
  - Type: `str`
  - Default: `""`
  - Description: Embedding API base URL

- `embedding_deployment_name`
  - Type: `str`
  - Default: `""`
  - Description: Embedding deployment name

- `embedding_model`
  - Type: `str`
  - Default: `"local"`
  - Description: Embedding model to use

**Message Handling**
- `max_message_chars`
  - Type: `int`

>   - Default: `10000`

Reviewer: ```suggestion
  - Default: `30000`
```

>   - Description: Maximum number of characters in an observation's content

Reviewer: ```suggestion
  - Description: The approximate maximum number of characters in the content of an event included in the prompt to the LLM. Larger observations are truncated.
```

**Retrying**
- `num_retries`
  - Type: `int`
  - Default: `8`
  - Description: Number of retries to attempt

- `retry_max_wait`
  - Type: `int`
  - Default: `120`
  - Description: Maximum wait time (in seconds) between retry attempts

- `retry_min_wait`
  - Type: `int`
  - Default: `15`
  - Description: Minimum wait time (in seconds) between retry attempts

- `retry_multiplier`
  - Type: `float`
  - Default: `2.0`
  - Description: Multiplier for exponential backoff calculation

**Advanced Options**
- `drop_params`
  - Type: `bool`
  - Default: `false`
  - Description: Drop any unmapped (unsupported) params without causing an exception

- `caching_prompt`
  - Type: `bool`
  - Default: `true`
  - Description: Using the prompt caching feature if provided by the LLM and supported

- `ollama_base_url`
  - Type: `str`
  - Default: `""`
  - Description: Base URL for the OLLAMA API

- `temperature`
  - Type: `float`
  - Default: `0.0`
  - Description: Temperature for the API

- `timeout`
  - Type: `int`
  - Default: `0`
  - Description: Timeout for the API

- `top_p`
  - Type: `float`
  - Default: `1.0`
  - Description: Top p for the API

- `disable_vision`
  - Type: `bool`

>   - Default: `true`

Reviewer: ```suggestion
  - Default: `None`
```
